File: sdks/go/http_client.go
package decktools

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"mime"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

type httpClient struct {
	root              string
	token             string
	apiKey            string
	spaceID           string
	spaceIDExplicit   bool
	authUUID          string
	client            *http.Client
	onUnauthorized    func(context.Context) (AuthRefresh, error)
	onPaymentRequired func(context.Context) error
	mu                sync.RWMutex
	authRefreshMu     sync.Mutex
	authRefreshWait   *authRefreshWaiter
	guestDowngradeMu  sync.Mutex
	guestDowngradeWait *guestDowngradeWaiter
	authDowngradedToGuest bool
}

type authRefreshWaiter struct {
	done chan struct{}
	auth AuthRefresh
	err  error
}

type guestDowngradeWaiter struct {
	done    chan struct{}
	spaceID string
	err     error
}

type httpResponse struct {
	Header http.Header
	Body   []byte
	Stream io.ReadCloser
}

func newHTTPClient(ctx context.Context, options ClientOptions) (*httpClient, error) {
	root := strings.TrimRight(options.Root, "/")
	if root == "" {
		root = DefaultRoot
	}
	authUUID, err := resolveAuthUUID(ctx, options)
	if err != nil {
		return nil, err
	}
	c := options.HTTPClient
	if c == nil {
		c = &http.Client{Timeout: 30 * time.Second}
	}
	return &httpClient{
		root:              root,
		token:             options.Token,
		apiKey:            options.APIKey,
		spaceID:           options.SpaceID,
		spaceIDExplicit:   options.SpaceID != "",
		authUUID:          authUUID,
		client:            c,
		onUnauthorized:    options.OnUnauthorized,
		onPaymentRequired: options.OnPaymentRequired,
	}, nil
}

func (c *httpClient) Root() string {
	return c.root
}

func (c *httpClient) SetToken(token string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.token = token
	if token != "" {
		c.authDowngradedToGuest = false
	}
	c.clearAutoResolvedSpaceID()
}

func (c *httpClient) SetAPIKey(apiKey string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.apiKey = apiKey
	if apiKey != "" {
		c.authDowngradedToGuest = false
	}
	c.clearAutoResolvedSpaceID()
}

func (c *httpClient) SetSpaceID(spaceID string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.spaceID = spaceID
	c.spaceIDExplicit = spaceID != ""
}

func (c *httpClient) SpaceID() string {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.spaceID
}

func (c *httpClient) ResolveSpaceID(ctx context.Context, spaceID string) (string, error) {
	if spaceID != "" {
		return spaceID, nil
	}

	c.mu.RLock()
	if c.spaceID != "" {
		id := c.spaceID
		c.mu.RUnlock()
		return id, nil
	}
	c.mu.RUnlock()

	// Resolve from GET /user. The endpoint only requires X-Auth-UUID, so
	// it works for both authenticated users and guests.
	var user UserSelf
	if _, err := c.getJSON(ctx, "/user", nil, nil, &user); err != nil {
		return "", err
	}
	if user.ID == "" {
		return "", fmt.Errorf("user.self did not return an id")
	}

	c.mu.Lock()
	if c.spaceID == "" {
		c.spaceID = user.ID
	}
	id := c.spaceID
	c.mu.Unlock()
	return id, nil
}

func (c *httpClient) clearAutoResolvedSpaceID() {
	if c.spaceIDExplicit {
		return
	}
	c.spaceID = ""
}

func (c *httpClient) AuthUUID() string {
	return c.authUUID
}

func (c *httpClient) getJSON(ctx context.Context, path string, query url.Values, headers http.Header, out any) (*httpResponse, error) {
	res, err := c.do(ctx, http.MethodGet, path, query, headers, nil, false)
	if err != nil {
		return nil, err
	}
	if err := decodeJSONBody(res.Body, out); err != nil {
		return nil, err
	}
	return res, nil
}

func (c *httpClient) postJSON(ctx context.Context, path string, in any, out any) (*httpResponse, error) {
	var body []byte
	var err error
	if in != nil {
		body, err = json.Marshal(in)
		if err != nil {
			return nil, err
		}
	}
	res, err := c.do(ctx, http.MethodPost, path, nil, nil, body, false)
	if err != nil {
		return nil, err
	}
	if err := decodeJSONBody(res.Body, out); err != nil {
		return nil, err
	}
	return res, nil
}

func (c *httpClient) postMultipart(ctx context.Context, path string, headers http.Header, body []byte, out any) (*httpResponse, error) {
	res, err := c.do(ctx, http.MethodPost, path, nil, headers, body, false)
	if err != nil {
		return nil, err
	}
	if err := decodeJSONBody(res.Body, out); err != nil {
		return nil, err
	}
	return res, nil
}

func (c *httpClient) putJSON(ctx context.Context, path string, query url.Values, in any, out any) (*httpResponse, error) {
	var body []byte
	var err error
	if in != nil {
		body, err = json.Marshal(in)
		if err != nil {
			return nil, err
		}
	}
	res, err := c.do(ctx, http.MethodPut, path, query, nil, body, false)
	if err != nil {
		return nil, err
	}
	if err := decodeJSONBody(res.Body, out); err != nil {
		return nil, err
	}
	return res, nil
}

// decodeJSONBody unmarshals JSON into out. Empty bodies (common for 204 / start)
// are treated as success and leave out unchanged.
func decodeJSONBody(body []byte, out any) error {
	if out == nil || len(bytes.TrimSpace(body)) == 0 {
		return nil
	}
	return json.Unmarshal(body, out)
}

func (c *httpClient) delete(ctx context.Context, path string, query url.Values) error {
	_, err := c.do(ctx, http.MethodDelete, path, query, nil, nil, false)
	return err
}

func (c *httpClient) eventStream(ctx context.Context, path string, query url.Values, headers http.Header) (*httpResponse, error) {
	return c.do(ctx, http.MethodGet, path, query, headers, nil, true)
}

func (c *httpClient) do(ctx context.Context, method string, path string, query url.Values, headers http.Header, body []byte, stream bool) (*httpResponse, error) {
	var paymentRetried, authRetried bool
	var lastErr error

	for attempt := 0; attempt <= len(retryDelays); attempt++ {
		if attempt > 0 {
			delay := retryDelayForAttempt(attempt - 1)
			if err := sleepContext(ctx, delay); err != nil {
				return nil, err
			}
		}

		res, err := c.doOnce(ctx, method, path, query, headers, body, stream)
		if err == nil {
			return res, nil
		}
		lastErr = err

		var apiErr *APIError
		if errors.As(err, &apiErr) {
			if apiErr.StatusCode == http.StatusPaymentRequired && !paymentRetried {
				if c.onPaymentRequired != nil {
					paymentRetried = true
					if err := c.onPaymentRequired(ctx); err != nil {
						return nil, err
					}
					attempt = -1
					continue
				}
				return nil, paymentRequiredError(apiErr)
			}
			if apiErr.StatusCode == http.StatusUnauthorized && !authRetried {
				authRetried = true
				oldSpaceID := spaceIDFromRequest(path, query, body)
				if oldSpaceID == "" {
					oldSpaceID = c.SpaceID()
				}

				if c.onUnauthorized != nil && c.hasUserToken() && !c.usesAPIKeyAuth() {
					auth, refreshErr := c.refreshAuth(ctx)
					if refreshErr == nil {
						c.SetToken(auth.Token)
						if auth.SpaceID != "" {
							c.SetSpaceID(auth.SpaceID)
						}
						path, body = rewriteSpaceID(path, query, body, oldSpaceID, auth.SpaceID)
						attempt = -1
						continue
					}
					// Refresh failed — fall through to guest mode.
				}

				if c.hasCredentials() || c.isAuthDowngradedToGuest() || c.hasGuestDowngradeInFlight() {
					guestSpaceID, guestErr := c.ensureGuestMode(ctx)
					if guestErr != nil {
						return nil, guestErr
					}
					path, body = rewriteSpaceID(path, query, body, oldSpaceID, guestSpaceID)
					attempt = -1
					continue
				}
				return nil, apiErr
			}
			if !isRetriableStatus(apiErr.StatusCode) {
				return nil, err
			}
			continue
		}

		if !isRetriableTransportError(err) || attempt >= len(retryDelays) {
			return nil, err
		}
	}
	return nil, lastErr
}

func (c *httpClient) usesAPIKeyAuth() bool {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.apiKey != "" && c.token == ""
}

func (c *httpClient) hasUserToken() bool {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.token != ""
}

func (c *httpClient) hasCredentials() bool {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.token != "" || c.apiKey != ""
}

func (c *httpClient) isAuthDowngradedToGuest() bool {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.authDowngradedToGuest
}

func (c *httpClient) hasGuestDowngradeInFlight() bool {
	c.guestDowngradeMu.Lock()
	defer c.guestDowngradeMu.Unlock()
	return c.guestDowngradeWait != nil
}

func (c *httpClient) refreshAuth(ctx context.Context) (AuthRefresh, error) {
	if c.onUnauthorized == nil {
		return AuthRefresh{}, fmt.Errorf("onUnauthorized is not configured")
	}

	c.authRefreshMu.Lock()
	if waiting := c.authRefreshWait; waiting != nil {
		c.authRefreshMu.Unlock()
		select {
		case <-waiting.done:
			return waiting.auth, waiting.err
		case <-ctx.Done():
			return AuthRefresh{}, ctx.Err()
		}
	}

	wait := &authRefreshWaiter{done: make(chan struct{})}
	c.authRefreshWait = wait
	c.authRefreshMu.Unlock()

	wait.auth, wait.err = c.onUnauthorized(ctx)

	c.authRefreshMu.Lock()
	if c.authRefreshWait == wait {
		c.authRefreshWait = nil
	}
	c.authRefreshMu.Unlock()
	close(wait.done)

	return wait.auth, wait.err
}

func (c *httpClient) ensureGuestMode(ctx context.Context) (string, error) {
	c.mu.RLock()
	alreadyGuest := c.authDowngradedToGuest && c.token == "" && c.apiKey == "" && c.spaceID != ""
	existingSpace := c.spaceID
	c.mu.RUnlock()

	c.guestDowngradeMu.Lock()
	inFlight := c.guestDowngradeWait != nil
	c.guestDowngradeMu.Unlock()

	if alreadyGuest && !inFlight {
		return existingSpace, nil
	}
	return c.downgradeToGuest(ctx)
}

func (c *httpClient) downgradeToGuest(ctx context.Context) (string, error) {
	c.guestDowngradeMu.Lock()
	if waiting := c.guestDowngradeWait; waiting != nil {
		c.guestDowngradeMu.Unlock()
		select {
		case <-waiting.done:
			return waiting.spaceID, waiting.err
		case <-ctx.Done():
			return "", ctx.Err()
		}
	}

	wait := &guestDowngradeWaiter{done: make(chan struct{})}
	c.guestDowngradeWait = wait
	c.guestDowngradeMu.Unlock()

	c.mu.Lock()
	c.authDowngradedToGuest = true
	c.token = ""
	c.apiKey = ""
	c.spaceID = ""
	c.spaceIDExplicit = false
	c.mu.Unlock()

	guestSpaceID, err := c.fetchUserSpaceID(ctx)
	if err == nil {
		c.mu.Lock()
		c.spaceID = guestSpaceID
		c.mu.Unlock()
	}

	wait.spaceID = guestSpaceID
	wait.err = err

	c.guestDowngradeMu.Lock()
	if c.guestDowngradeWait == wait {
		c.guestDowngradeWait = nil
	}
	c.guestDowngradeMu.Unlock()
	close(wait.done)

	return wait.spaceID, wait.err
}

// fetchUserSpaceID calls GET /user directly to avoid re-entering 401 guest handling.
func (c *httpClient) fetchUserSpaceID(ctx context.Context) (string, error) {
	res, err := c.doOnce(ctx, http.MethodGet, "/user", nil, nil, nil, false)
	if err != nil {
		return "", err
	}
	var user UserSelf
	if err := decodeJSONBody(res.Body, &user); err != nil {
		return "", err
	}
	if user.ID == "" {
		return "", fmt.Errorf("Failed to resolve guest space id after auth downgrade")
	}
	return user.ID, nil
}

func (c *httpClient) doOnce(ctx context.Context, method string, path string, query url.Values, headers http.Header, body []byte, stream bool) (*httpResponse, error) {
	u, err := url.Parse(c.root + "/" + strings.TrimLeft(path, "/"))
	if err != nil {
		return nil, err
	}
	if query != nil {
		u.RawQuery = query.Encode()
	}

	var reader io.Reader
	if body != nil {
		reader = bytes.NewReader(body)
	}
	req, err := http.NewRequestWithContext(ctx, method, u.String(), reader)
	if err != nil {
		return nil, err
	}
	c.applyHeaders(req.Header)
	for key, values := range headers {
		req.Header.Del(key)
		for _, value := range values {
			req.Header.Add(key, value)
		}
	}
	if body != nil && req.Header.Get("Content-Type") == "" {
		req.Header.Set("Content-Type", "application/json")
	}

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		defer resp.Body.Close()
		data, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
		return nil, newAPIError(resp, data)
	}
	if stream {
		return &httpResponse{Header: resp.Header.Clone(), Stream: resp.Body}, nil
	}
	defer resp.Body.Close()
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	return &httpResponse{Header: resp.Header.Clone(), Body: data}, nil
}

func (c *httpClient) applyHeaders(headers http.Header) {
	c.mu.RLock()
	defer c.mu.RUnlock()
	headers.Set("Content-Type", "application/json")
	headers.Set("X-Auth-UUID", c.authUUID)
	if c.token != "" {
		headers.Set("X-Auth-Token", c.token)
	}
	if c.apiKey != "" {
		headers.Set("Authorization", "Bearer "+c.apiKey)
	}
}

func sleepContext(ctx context.Context, delay time.Duration) error {
	timer := time.NewTimer(delay)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-timer.C:
		return nil
	}
}

func spaceIDFromRequest(path string, query url.Values, body []byte) string {
	if query != nil {
		if spaceID := query.Get("spaceId"); spaceID != "" {
			return spaceID
		}
	}
	if len(body) > 0 {
		var payload map[string]any
		if json.Unmarshal(body, &payload) == nil {
			if spaceID, ok := payload["spaceId"].(string); ok && spaceID != "" {
				return spaceID
			}
		}
	}
	if marker := "/spaces/"; strings.Contains(path, marker) {
		rest := path[strings.Index(path, marker)+len(marker):]
		if end := strings.IndexByte(rest, '/'); end > 0 {
			spaceID, err := url.PathUnescape(rest[:end])
			if err == nil && spaceID != "" {
				return spaceID
			}
			return rest[:end]
		}
	}
	return ""
}

func rewriteSpaceID(path string, query url.Values, body []byte, oldSpaceID string, newSpaceID string) (string, []byte) {
	if oldSpaceID == "" || newSpaceID == "" || oldSpaceID == newSpaceID {
		return path, body
	}
	escapedOld := url.PathEscape(oldSpaceID)
	escapedNew := url.PathEscape(newSpaceID)
	path = strings.ReplaceAll(path, "/spaces/"+escapedOld+"/", "/spaces/"+escapedNew+"/")
	path = strings.ReplaceAll(path, "/spaces/"+oldSpaceID+"/", "/spaces/"+newSpaceID+"/")
	if query != nil && query.Get("spaceId") == oldSpaceID {
		query.Set("spaceId", newSpaceID)
	}
	if len(body) == 0 {
		return path, body
	}
	var payload map[string]any
	if json.Unmarshal(body, &payload) == nil && payload["spaceId"] == oldSpaceID {
		payload["spaceId"] = newSpaceID
		next, err := json.Marshal(payload)
		if err == nil {
			return path, next
		}
	}
	return path, body
}

func isEventStream(headers http.Header) bool {
	contentType := strings.ToLower(headers.Get("Content-Type"))
	mediaType, _, err := mime.ParseMediaType(contentType)
	if err == nil {
		contentType = mediaType
	}
	return strings.Contains(contentType, "event-stream")
}

File: sdks/python/decktools/http_client.py
from __future__ import annotations

import contextlib
import re
import threading
import time
from collections.abc import Callable, Iterator, Mapping
from typing import Any
from urllib.parse import unquote

import httpx

from .auth_uuid import resolve_auth_uuid
from .errors import RETRIABLE_STATUS_CODES, RETRY_DELAYS, APIError, is_retriable_exception
from .types import DEFAULT_ROOT, AuthUUIDStorage

AuthRefresh = str | Mapping[str, str]

_SPACE_PATH_RE = re.compile(r"/spaces/([^/?#]+)/")


class HTTPClient:
    def __init__(
        self,
        *,
        root: str = DEFAULT_ROOT,
        token: str | None = None,
        api_key: str | None = None,
        space_id: str | None = None,
        auth_uuid: str | None = None,
        auth_uuid_storage: AuthUUIDStorage | None = None,
        on_unauthorized: Callable[[], AuthRefresh] | None = None,
        on_payment_required: Callable[[], None] | None = None,
        client: httpx.Client | None = None,
    ) -> None:
        self.root = (root or DEFAULT_ROOT).rstrip("/")
        self.token = token
        self.api_key = api_key
        self.space_id = space_id
        self._space_id_explicit = bool(space_id)
        self.auth_uuid = resolve_auth_uuid(auth_uuid=auth_uuid, storage=auth_uuid_storage)
        self.on_unauthorized = on_unauthorized
        self.on_payment_required = on_payment_required
        self.client = client or httpx.Client(timeout=30.0)
        self._owns_client = client is None
        self._state_lock = threading.RLock()
        self._refresh_lock = threading.Lock()
        self._guest_downgrade_lock = threading.Lock()
        self._guest_downgrade_waiters: list[threading.Event] = []
        self._guest_downgrade_error: BaseException | None = None
        self._guest_downgrade_in_flight = False
        self._auth_downgraded_to_guest = False

    def close(self) -> None:
        if self._owns_client:
            self.client.close()

    def set_token(self, token: str | None) -> None:
        with self._state_lock:
            self.token = token
            if token:
                self._auth_downgraded_to_guest = False
            self._clear_auto_space_id()

    def set_api_key(self, api_key: str | None) -> None:
        with self._state_lock:
            self.api_key = api_key
            if api_key:
                self._auth_downgraded_to_guest = False
            self._clear_auto_space_id()

    def set_space_id(self, space_id: str | None) -> None:
        with self._state_lock:
            self.space_id = space_id
            self._space_id_explicit = bool(space_id)

    def _clear_auto_space_id(self) -> None:
        if not self._space_id_explicit:
            self.space_id = None

    def resolve_space_id(self, space_id: str | None = None) -> str:
        if space_id:
            return space_id
        with self._state_lock:
            if self.space_id:
                return self.space_id

        # Resolve from GET /user. Only X-Auth-UUID is required, so this works
        # for authenticated users and guests.
        response = self.request("GET", "/user")
        user = response.json()
        resolved = user.get("id") if isinstance(user, Mapping) else None
        if not isinstance(resolved, str) or not resolved:
            raise ValueError("user.self did not return an id")
        with self._state_lock:
            if not self.space_id:
                self.space_id = resolved
            return self.space_id

    def url(self, path: str) -> str:
        return f"{self.root}/{path.lstrip('/')}"

    def _auth_headers(self) -> dict[str, str]:
        headers = {"Content-Type": "application/json", "X-Auth-UUID": self.auth_uuid}
        with self._state_lock:
            if self.token:
                headers["X-Auth-Token"] = self.token
            if self.api_key:
                headers["Authorization"] = f"Bearer {self.api_key}"
        return headers

    def _is_api_key_auth(self) -> bool:
        with self._state_lock:
            return bool(self.api_key and not self.token)

    def _has_credentials(self) -> bool:
        with self._state_lock:
            return bool(self.token or self.api_key)

    def _refresh(self, request_token: str | None) -> tuple[str, str | None]:
        if self.on_unauthorized is None:
            raise RuntimeError("on_unauthorized is not configured")
        with self._refresh_lock:
            with self._state_lock:
                if self.token != request_token and self.token:
                    return self.token, self.space_id
            result = self.on_unauthorized()
            if isinstance(result, str):
                next_token = result
                next_space = self.space_id
            else:
                next_token = str(result.get("token") or "")
                next_space = result.get("spaceId") or result.get("space_id") or self.space_id
            if not next_token:
                raise ValueError("on_unauthorized did not return a token")
            self.set_token(next_token)
            if next_space:
                self.set_space_id(next_space)
            return next_token, next_space

    def _ensure_guest_mode(self) -> str:
        with self._state_lock:
            if (
                self._auth_downgraded_to_guest
                and not self.token
                and not self.api_key
                and self.space_id
                and not self._guest_downgrade_in_flight
            ):
                return self.space_id

        with self._guest_downgrade_lock:
            if self._guest_downgrade_in_flight:
                waiter = threading.Event()
                self._guest_downgrade_waiters.append(waiter)
            else:
                waiter = None
                self._guest_downgrade_in_flight = True
                self._guest_downgrade_error = None
                with self._state_lock:
                    self._auth_downgraded_to_guest = True
                    self.token = None
                    self.api_key = None
                    self.space_id = None
                    self._space_id_explicit = False

        if waiter is not None:
            waiter.wait()
            if self._guest_downgrade_error is not None:
                raise self._guest_downgrade_error
            with self._state_lock:
                if not self.space_id:
                    raise RuntimeError("Failed to resolve guest space id after auth downgrade")
                return self.space_id

        try:
            # Bypass request() so a nested 401 cannot re-enter guest downgrade.
            response = self.client.request(
                "GET",
                self.url("/user"),
                headers=self._auth_headers(),
            )
            if not response.is_success:
                raise APIError.from_response(response)
            user = response.json()
            guest_space = user.get("id") if isinstance(user, Mapping) else None
            if not isinstance(guest_space, str) or not guest_space:
                raise RuntimeError("Failed to resolve guest space id after auth downgrade")
            with self._state_lock:
                self.space_id = guest_space
            return guest_space
        except BaseException as error:
            self._guest_downgrade_error = error
            raise
        finally:
            with self._guest_downgrade_lock:
                self._guest_downgrade_in_flight = False
                waiters = list(self._guest_downgrade_waiters)
                self._guest_downgrade_waiters.clear()
            for pending in waiters:
                pending.set()

    @staticmethod
    def _space_id_from_request(
        *,
        url: str,
        params: dict[str, Any] | None,
        json_body: Any,
    ) -> str | None:
        if params and isinstance(params.get("spaceId"), str):
            return params["spaceId"]
        if isinstance(json_body, Mapping) and isinstance(json_body.get("spaceId"), str):
            return json_body["spaceId"]
        match = _SPACE_PATH_RE.search(url)
        if match:
            return unquote(match.group(1))
        return None

    @staticmethod
    def _rewrite_space(
        *,
        url: str,
        params: dict[str, Any] | None,
        json_body: Any,
        old_space: str | None,
        new_space: str | None,
    ) -> tuple[str, dict[str, Any] | None, Any]:
        if not old_space or not new_space or old_space == new_space:
            return url, params, json_body
        url = url.replace(f"/spaces/{old_space}/", f"/spaces/{new_space}/")
        if params and params.get("spaceId") == old_space:
            params = {**params, "spaceId": new_space}
        if isinstance(json_body, Mapping) and json_body.get("spaceId") == old_space:
            json_body = {**json_body, "spaceId": new_space}
        return url, params, json_body

    def _handle_unauthorized(
        self,
        *,
        request_token: str | None,
        url: str,
        mutable_params: dict[str, Any] | None,
        json_body: Any,
    ) -> tuple[str, dict[str, Any] | None, Any] | None:
        """Try token refresh, then guest downgrade. Returns rewritten request parts, or None to raise."""
        old_space = self._space_id_from_request(
            url=url, params=mutable_params, json_body=json_body
        ) or self.space_id

        if self.on_unauthorized is not None and self.token and not self._is_api_key_auth():
            try:
                _, next_space = self._refresh(request_token)
                return self._rewrite_space(
                    url=url,
                    params=mutable_params,
                    json_body=json_body,
                    old_space=old_space,
                    new_space=next_space,
                )
            except Exception:
                # Refresh failed — fall through to guest mode.
                pass

        with self._state_lock:
            can_downgrade = bool(
                self.token
                or self.api_key
                or self._auth_downgraded_to_guest
                or self._guest_downgrade_in_flight
            )
        if can_downgrade:
            guest_space = self._ensure_guest_mode()
            return self._rewrite_space(
                url=url,
                params=mutable_params,
                json_body=json_body,
                old_space=old_space,
                new_space=guest_space,
            )
        return None

    def request(
        self,
        method: str,
        path: str,
        *,
        params: Mapping[str, Any] | None = None,
        json: Any = None,
        data: Mapping[str, Any] | None = None,
        files: Any = None,
        headers: Mapping[str, str] | None = None,
    ) -> httpx.Response:
        url = self.url(path)
        mutable_params = dict(params) if params else None
        json_body = json
        form_data = dict(data) if data else None
        auth_retried = False
        payment_retried = False
        network_attempt = 0

        while True:
            if network_attempt:
                time.sleep(RETRY_DELAYS[min(network_attempt - 1, len(RETRY_DELAYS) - 1)])
            request_token = self.token
            merged_headers = {**self._auth_headers(), **dict(headers or {})}
            if files is not None or form_data is not None:
                merged_headers.pop("Content-Type", None)
            try:
                response = self.client.request(
                    method,
                    url,
                    params=mutable_params,
                    json=json_body,
                    data=form_data,
                    files=files,
                    headers=merged_headers,
                )
            except Exception as error:
                if (
                    isinstance(error, Exception)
                    and is_retriable_exception(error)
                    and network_attempt < len(RETRY_DELAYS)
                ):
                    network_attempt += 1
                    continue
                raise

            if response.is_success:
                return response

            base = APIError.from_response(response)
            if response.status_code == 402 and not payment_retried:
                if self.on_payment_required is None:
                    raise APIError.payment_required(base)
                payment_retried = True
                self.on_payment_required()
                network_attempt = 0
                continue

            if response.status_code == 401 and not auth_retried:
                auth_retried = True
                rewritten = self._handle_unauthorized(
                    request_token=request_token,
                    url=url,
                    mutable_params=mutable_params,
                    json_body=json_body if json_body is not None else form_data,
                )
                if rewritten is not None:
                    url, mutable_params, rewritten_body = rewritten
                    if json_body is not None:
                        json_body = rewritten_body
                    elif form_data is not None and isinstance(rewritten_body, dict):
                        form_data = rewritten_body
                    network_attempt = 0
                    continue
                raise base

            if response.status_code in RETRIABLE_STATUS_CODES and network_attempt < len(
                RETRY_DELAYS
            ):
                network_attempt += 1
                continue
            raise base

    @contextlib.contextmanager
    def stream(
        self,
        path: str,
        *,
        params: Mapping[str, Any] | None = None,
        headers: Mapping[str, str] | None = None,
    ) -> Iterator[httpx.Response]:
        url = self.url(path)
        mutable_params = dict(params) if params else None
        auth_retried = False
        payment_retried = False
        network_attempt = 0

        while True:
            if network_attempt:
                time.sleep(RETRY_DELAYS[min(network_attempt - 1, len(RETRY_DELAYS) - 1)])
            request_token = self.token
            manager = self.client.stream(
                "GET",
                url,
                params=mutable_params,
                headers={**self._auth_headers(), **dict(headers or {})},
            )
            try:
                response = manager.__enter__()
            except Exception as error:
                if is_retriable_exception(error) and network_attempt < len(RETRY_DELAYS):
                    network_attempt += 1
                    continue
                raise

            if response.is_success:
                try:
                    yield response
                finally:
                    manager.__exit__(None, None, None)
                return

            response.read()
            base = APIError.from_response(response)
            manager.__exit__(None, None, None)
            if response.status_code == 402 and not payment_retried:
                if self.on_payment_required is None:
                    raise APIError.payment_required(base)
                payment_retried = True
                self.on_payment_required()
                network_attempt = 0
                continue
            if response.status_code == 401 and not auth_retried:
                auth_retried = True
                rewritten = self._handle_unauthorized(
                    request_token=request_token,
                    url=url,
                    mutable_params=mutable_params,
                    json_body=None,
                )
                if rewritten is not None:
                    url, mutable_params, _ = rewritten
                    network_attempt = 0
                    continue
                raise base
            if response.status_code in RETRIABLE_STATUS_CODES and network_attempt < len(
                RETRY_DELAYS
            ):
                network_attempt += 1
                continue
            raise base

    # TypeScript-compatible aliases.
    setToken = set_token
    setApiKey = set_api_key
    setSpaceId = set_space_id
    resolveSpaceId = resolve_space_id
